arangod/RocksDBEngine/RocksDBPrimaryIndex.cpp

```diff
@@ -425,42 +550,224 @@ IndexIterator* RocksDBPrimaryIndex::iteratorForCondition(
+          type == aql::NODE_TYPE_OPERATOR_BINARY_GT ||
+          type == aql::NODE_TYPE_OPERATOR_BINARY_EQ
+        )) {
+      // We should throw to show that the condition is broken
+      // LOG_DEVEL << "compare not in in < > <= =>";
+      //comp->dump(20);
+      return new EmptyIndexIterator(&_collection, trx);
+    }
+
+    auto attrNode = comp->getMember(0);
+    auto valNode = comp->getMember(1);
+    bool flip = false;
+
+    if (attrNode->type != aql::NODE_TYPE_ATTRIBUTE_ACCESS) {
+      // value == a.b  ->  flip the two sides
+      // REVIEW - taken from VPackIndex -> impact on comp?! this is not
+      attrNode = comp->getMember(1);
+      valNode = comp->getMember(0);
+      flip = true;
+    }
+
+    TRI_ASSERT(attrNode->type == aql::NODE_TYPE_ATTRIBUTE_ACCESS);
+    bool const isId = (attrNode->stringEquals(StaticStrings::IdString));
+
+    std::string value = lowest;
+    if (valNode->isStringValue()) {
+      value = valNode->getString();
+    } else if (valNode->isObject() || valNode->isArray()) {
+      value = hightest;
+    } else if (valNode->isNullValue() || valNode->isBoolValue()) {
+      // keep lowest
+    } else {
+      TRI_ASSERT(false);
     }
+
+    if (flip) {
+      // LOG_DEVEL << "flipping";
+      switch (type) {
+        case aql::NODE_TYPE_OPERATOR_BINARY_LE: {
+          type = aql::NODE_TYPE_OPERATOR_BINARY_GE;
+          break;
+        }
+        case aql::NODE_TYPE_OPERATOR_BINARY_LT: {
+          type = aql::NODE_TYPE_OPERATOR_BINARY_GT;
+          break;
+        }
+        case aql::NODE_TYPE_OPERATOR_BINARY_GE: {
+          type = aql::NODE_TYPE_OPERATOR_BINARY_LE;
+          break;
+        }
+        case aql::NODE_TYPE_OPERATOR_BINARY_GT: {
+          type = aql::NODE_TYPE_OPERATOR_BINARY_LT;
+          break;
+        }
+        case aql::NODE_TYPE_OPERATOR_BINARY_EQ: {
+          break;
+        }
+        default: {
+          continue;
+          // LOG_DEVEL << "cound not specialize condition";
+          // LOG_DEVEL << "EmptyIndexIterator - complex"
+          //          << arangodb::aql::AstNode::toString(node);
+          // return new EmptyIndexIterator(&_collection, trx);
+        }
+      }
+    }
+
+    if (!removeCollectionFromString(isId, value)) {
+      // _id is not valid for this collection skipping - THROW ERROR?
+      continue;
+    }
+
+    if (type == aql::NODE_TYPE_OPERATOR_BINARY_EQ) {
+      // a.b < value
+      if (!upper || value < *upper) {
+        upper = std::make_unique<std::string>(value);
+      }
+      if (!lower || value < *lower) {
+        upper = std::make_unique<std::string>(std::move(value));
+      }
+    }
+
+    if (type == aql::NODE_TYPE_OPERATOR_BINARY_LE || type == aql::NODE_TYPE_OPERATOR_BINARY_LT) {
+      // a.b < value
+      if (type == aql::NODE_TYPE_OPERATOR_BINARY_LT && value != lowest) {
+        // LOG_DEVEL << "upper pre modification" << value;
+        value.back() -= 0x01U;  // modify upper bound so that it is not included
```

Comment:
what if this underflows?

Reply:
what if `value` is the empty string?

Reply:
Expecting valid `_key` and `_id` values.